text/0018-message-ordering-guarantee.md

```diff
@@ -0,0 +1,171 @@
+
+Likewise, the same is true of internal commands:
+
+```C#
+GameObject m_obj;
+
+void Update()
+{
+    m_obj = GameObject.Instantiate(prefab, location, rotation);
+    NetworkBehavior networkBehavior = m_obj.GetComponent<NetworkBehavior>();
+    networkBehavior.Spawn();
+}
+
+void LateUpdate()
+{
+    ulong newOwnerClientId = GetNewOwnerClientId();
+    NetworkBehavior networkBehavior = m_obj.GetComponent<NetworkBehavior>();
+    networkBehavior.ChangeOwnership(newOwnerClientId);
+}
+```
+
+In this example, because `Spawn()` is called during `Update()` on the server, and `ChangeOwnership()` is called during `LateUpdate()` on the server, the same will be true on the client - the object will be spawned on the client during the `Update` stage, and its ownership changed during `LateUpdate`.
+
+# Reference-level explanation
+
+There are two main refactors involved in implementing this change:
+
+1. The first and largest refactor is a change in the way internal messages are sent to ensure they use the same mechanism as RPCs, and the corresponding refactor of that mechanism to ensure it's sufficiently generic to support all message types.
+
+2. The second, smaller refactor is changing the default value of the UpdateStage value in the send parameters so that it reads from the current update stage rather than defaulting to 0 (Update).
+
+The vast majority of the work is in the former refactor, but the biggest user-facing change is the latter one.
+
+In order to complete this refactor, a few things need to happen:
+
+- All RPC-related classes will need to be renamed from Rpc\* to Message\*. This accounts for RpcFrameQueueItem, RpcQueueContainer, RpcQueueProcessor, RpcBatcher, and RpcQueueHistoryFrame.
+
+- All internal commands must be sent through `RpcQueueProcessor.SendFrameQueueItem()`. As a consequence, this also means all internal commands must now build `RpcFrameQueueItem` structs.
+
+- `RpcQueueContainer.QueueItemType` now contains only three values: `ClientRpc`, `ServerRpc`, and `InternalCommand`. `CreateObject` and `DestroyObject` are removed, and are now a subcategory of `InternalCommand` <sup>[See Question #1]</sup>
+
+- `MessageQueueProcessor`'s (formerly `RpcQueueProcessor`) `m_InternalMLAPISendQueue` is removed, along with all functions in `MessageQueueProcessor` and `MessageQueueContainer` that reference it.
+
+- On the receiving end, the handling of all internal messages using the previous values from `NetworkConstants.cs` will instead handle a new `INTERNAL_COMMAND` constant, as it will need to pull the encapsulated message from the queue and process it before passing the wrapped content to the existing logic
+
+- Since structs can't have parameterless constructors, `NetworkUpdateStage` will be adjusted to appear as follows:
+  
+  ```C#
+  public enum NetworkUpdateStage : byte  
```

Comment:
Could you expand upon why something like this:
```
    public enum NetworkUpdateStage : byte
    {
        Initialization = 1,
        EarlyUpdate = 2,
        FixedUpdate = 3,
        PreUpdate = 4,
        Update = 0, // Default
        PreLateUpdate = 5,
        PostLateUpdate = 6
    }
```
 Would be different (as far as structs are concerned) than what you proposed and how adding the Unset value solves for the issue?  The original enum still has a "zero value" (i.e. default if not set), and so this area could potentially use some additional language as to why we need to make this change. 
(_mostly out of curiosity, but also to better clarify the justification_)

Reply:
It's different because if Update is the 0 value, we wouldn't be able to tell the difference between "user didn't specify" (meaning we would have to do the smart logic to use the current stage) or "user explicitly requested this to run at the Update stage" (in which case we must not change it). I will update the RFC to be clearer on that.

Reply:
Updated

Reply:
+1 on this. if we'll provide a more intelligent default in the absence of a specific one, that'd be `Unset` not `Default` because we couldn't determine if it was really set or not otherwise. lgtm!